lib/trento/software_updates/discovery.ex

```diff
@@ -13,5 +21,90 @@ defmodule Trento.SoftwareUpdates.Discovery do
   def get_relevant_patches(system_id),
     do: adapter().get_relevant_patches(system_id)
 
+  @spec discover_software_updates :: {:ok, {list(), list()}}
+  def discover_software_updates,
+    do:
+      {:ok,
+       Hosts.get_all_hosts()
+       |> Enum.map(&discover_host_software_updates/1)
+       |> Enum.split_with(fn
+         {:ok, _, :discovered, _, _} -> true
+         _ -> false
+       end)}
+
+  defp discover_host_software_updates(%HostReadModel{
+         id: host_id,
+         fully_qualified_domain_name: nil
+       }) do
+    Logger.info("Host #{host_id} does not have an fqdn. Skipping software updates discovery")
+    {:error, host_id, :host_without_fqdn}
+  end
+
+  defp discover_host_software_updates(%HostReadModel{
+         id: host_id,
+         fully_qualified_domain_name: fully_qualified_domain_name
+       }) do
+    with {:ok, system_id} <- get_system_id(fully_qualified_domain_name),
+         {:ok, relevant_patches} <- get_relevant_patches(system_id),
+         :ok <-
+           host_id
+           |> build_discovery_completion_command(relevant_patches)
+           |> commanded().dispatch() do
+      {:ok, host_id, :discovered, system_id, relevant_patches}
+    else
+      error ->
+        Logger.error(
+          "An error occurred during software updates discovery for host #{host_id}:  #{inspect(error)}"
+        )
+
+        {:error, host_id, error}
+    end
+  end
+
+  defp build_discovery_completion_command(host_id, relevant_patches),
+    do:
+      CompleteSoftwareUpdatesDiscovery.new!(%{
+        host_id: host_id,
+        relevant_patches:
+          Enum.reduce(
+            relevant_patches,
+            %{
+              security_advisories: 0,
+              bug_fixes: 0,
+              software_enhancements: 0
+            },
+            &track_relevant_patches/2
+          )
+      })
+
+  defp track_relevant_patches(%{advisory_type: advisory_type}, %{
+         security_advisories: security_advisories,
+         bug_fixes: bug_fixes,
+         software_enhancements: software_enhancements
+       }) do
+    advisory_type = AdvisoryType.from_string(advisory_type)
+
+    %{
+      security_advisories: increment_security_advisories(advisory_type, security_advisories),
```

Comment:
Is there a reason to abstract this into a function? Also why does the function take an `AdvisoryType`?

Reply:
Addressed differently here https://github.com/trento-project/web/pull/2376/commits/dcf5f14774bd12c7a19e0f445ab2e3432b920e0e

If it helps, I am happy :smile: